base/mmap.jl

```diff
@@ -1,168 +1,175 @@
-    if (filelen < offset + len)
-        systemerror("pwrite", ccall(:jl_pwrite, Cssize_t, (Cint, Ptr{Void}, UInt, FileOffset), fd, Int8[0], 1, offset + len - 1) < 1)
-    end
-    cpos = ccall(:jl_lseek, FileOffset, (Cint, FileOffset, Cint), fd, cpos, SEEK_SET)
-    systemerror("lseek", cpos < 0)
-    return Libc.mmap(len, prot, flags, fd, offset)
-end
+const PROT_READ     = convert(Cint,1)
+const PROT_WRITE    = convert(Cint,2)
+const MAP_SHARED    = convert(Cint,1)
+const MAP_PRIVATE   = convert(Cint,2)
+const MAP_ANONYMOUS = convert(Cint, @osx? 0x1000 : 0x20)
+const F_GETFL       = convert(Cint,3)
 
-# Determine a stream's read/write mode, and return prot & flags
-# appropriate for mmap
-# We could use isreadonly here, but it's worth checking that it's readable too
-function mmap_stream_settings(s::IO)
-    const PROT_READ::Cint = 1
-    const PROT_WRITE::Cint = 2
-    const MAP_SHARED::Cint = 1
-    const F_GETFL::Cint = 3
-    mode = ccall(:fcntl,Cint,(Cint,Cint),fd(s),F_GETFL)
-    systemerror("fcntl F_GETFL", mode == -1)
-    mode = mode & 3
-    if mode == 0
-        prot = PROT_READ
-    elseif mode == 1
-        prot = PROT_WRITE
-    else
+gethandle(io::IO) = fd(io)
+
+# Determine a stream's read/write mode, and return prot & flags appropriate for mmap
+function settings(s::Int, shared::Bool)
+    flags = shared ? MAP_SHARED : MAP_PRIVATE
+    if s == INVALID_HANDLE_VALUE
+        flags |= MAP_ANONYMOUS
         prot = PROT_READ | PROT_WRITE
+    else
+        mode = ccall(:fcntl,Cint,(Cint,Cint),s,F_GETFL)
+        systemerror("fcntl F_GETFL", mode == -1)
+        mode = mode & 3
+        prot = mode == 0 ? PROT_READ : mode == 1 ? PROT_WRITE : PROT_READ | PROT_WRITE
+        if prot & PROT_READ == 0
+            throw(ArgumentError("mmap requires read permissions on the file (choose r+)"))
+        end
     end
-    if prot & PROT_READ == 0
-        throw(ArgumentError("mmap requires read permissions on the file (choose r+)"))
-    end
-    flags = MAP_SHARED
     return prot, flags, (prot & PROT_WRITE) > 0
 end
 
-# Mmapped-array constructor
-function mmap_array{T,N}(::Type{T}, dims::NTuple{N,Integer}, s::IO, offset::FileOffset; grow::Bool=true)
-    prot, flags, iswrite = mmap_stream_settings(s)
-    len = prod(dims)*sizeof(T)
-    if len > typemax(Int)
-        throw(ArgumentError("file is too large to memory-map on this platform"))
-    end
-    if iswrite && grow
-        pmap, delta = mmap_grow(len, prot, flags, fd(s), offset)
-    else
-        pmap, delta = Libc.mmap(len, prot, flags, fd(s), offset)
+# Before mapping, grow the file to sufficient size
+# Note: a few mappable streams do not support lseek. When Julia
+# supports structures in ccall, switch to fstat.
+grow!(::Anonymous,o::Integer,l::Integer) = return
+function grow!(io::IO, offset::Integer, len::Integer)
+    pos = position(io)
+    filelen = filesize(io)
+    if filelen < offset + len
+        write(io, Base.zeros(UInt8,(offset + len) - filelen))
```

Comment:
This should probably go back to using `pwrite`, since that only had to write a single byte, whereas this is allocating and writing a potentially large array of zeros. It also doesn't seem that this seeks to the end of the file first, so it doesn't seem like it would yield the correct behavior if the file already exists?